### PR TITLE
Split abduction solver from SmtEngine

### PR DESCRIPTION
**Pull request by ajreynol**
This splits everything related to abducts into its own standalone module, AbductionSolver.

It furthermore converts some of the interfaces of SmtEngine to make them take Node instead of Expr (this will be done for every method eventually).

The code for interpolation should follow a similar pattern, e.g. InterpolantSolver.

### CODE REVIEW

File: src/smt/abduction_solver.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  /**
+   * Check that a solution to an abduction conjecture is indeed a solution.
+   *
+   * The check is made by determining that the assertions conjoined with the
+   * solution to the abduction problem (a) is SAT, and the assertions conjoined
+   * with the abduct and the goal is UNSAT. If these criteria are not met, an
+   * internal error is thrown.
+   */
+  void checkAbduct(Node a);
+ private:
```
Comment (Ying1123) [inline src/smt/abduction_solver.h:67]: empty line before this line.

File: src/smt/command.cpp [hunk 2]
```diff
@@ -1623,9 +1624,10 @@ void GetValueCommand::invoke(SmtEngine* smtEngine)
     for (int i = 0, size = d_terms.size(); i < size; i++)
     {
       Expr e = d_terms[i];
+      Node en = Node::fromExpr(e);
```
Comment (Ying1123) [inline src/smt/command.cpp:1627]: I think name like eNode will be more readable, but depends on you.

File: src/smt/command.cpp [hunk 3]
```diff
@@ -2199,15 +2201,19 @@ void GetAbductCommand::invoke(SmtEngine* smtEngine)
 {
   try
   {
+    Node conjn = Node::fromExpr(d_conj);
```
Comment (Ying1123) [inline src/smt/command.cpp:2204]: same here

File: src/smt/smt_engine.cpp [hunk 4]
```diff
@@ -86,6 +86,7 @@
 #include "smt/defined_function.h"
 #include "smt/logic_request.h"
 #include "smt/managed_ostreams.h"
+#include "smt/abduction_solver.h"
```
Comment (Ying1123) [inline src/smt/smt_engine.cpp:89]: This should be sorted in dictionary order.

File: src/smt/smt_engine.cpp [hunk 5]
```diff
@@ -94,6 +95,7 @@
 #include "smt/smt_engine_stats.h"
 #include "smt/term_formula_removal.h"
 #include "smt/update_ostream.h"
+#include "smt/abduction_solver.h"
```
Comment (Ying1123) [inline src/smt/smt_engine.cpp:98]: same here

File: src/smt/abduction_solver.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+#include "theory/quantifiers/sygus/sygus_grammar_cons.h"
+
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace smt {
+  
+AbductionSolver::AbductionSolver(SmtEngine * parent) : d_parent(parent)
+{
+
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/abduction_solver.cpp]: I've clang-formatted now.

File: src/smt/smt_engine.cpp (hunk 4 above)
Comment (ajreynol) [reply src/smt/smt_engine.cpp]: I've clang-formatted now.

File: src/smt/abduction_solver.cpp [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace smt {
+
+AbductionSolver::AbductionSolver(SmtEngine* parent) : d_parent(parent) {}
+
+AbductionSolver::~AbductionSolver() {}
+bool AbductionSolver::getAbduct(const Node& conj,
```
Comment (Ying1123) [inline src/smt/abduction_solver.cpp:32]: empty line before this line?

File: src/smt/abduction_solver.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+   *
+   * The argument grammarType is a sygus datatype type that encodes the syntax
+   * restrictions on the shape of possible solutions.
+   *
+   * This method invokes a separate copy of the SMT engine for solving the
+   * corresponding sygus problem for generating such a solution.
+   */
+  bool getAbduct(const Node& conj, const TypeNode& grammarType, Node& abd);
+
+  /** Same as above, but without user-provided grammar restrictions */
```
Comment (yoni206) [inline src/smt/abduction_solver.h:55]: Maybe mention that a default grammar is chosen internally?

File: src/smt/smt_engine.cpp [hunk 9]
```diff
@@ -81,6 +81,7 @@
 #include "proof/theory_proof.h"
 #include "proof/unsat_core.h"
 #include "prop/prop_engine.h"
+#include "smt/abduction_solver.h"
```
Comment (yoni206) [inline src/smt/smt_engine.cpp:84]: do you still need to include `sygus_abduct.h` (few rows down)?

File: src/smt/smt_engine.h [hunk 10]
```diff
@@ -508,7 +510,7 @@ class CVC4_PUBLIC SmtEngine
    *
    * @throw TypeCheckingException, LogicException, UnsafeInterruptException
    */
-  Expr expandDefinitions(const Expr& e);
+  Node expandDefinitions(const Node& e);
```
Comment (yoni206) [inline src/smt/smt_engine.h:513]: I am surprised that this particular change does not propagate to many more places that are unrelated to abduction. It might be a good idea to separate this particular change (and its propagated changes) to a separate PR. But if it is pretty local to abduction, then I guess it is fine to "merge" these two changes together.

File: src/smt/smt_engine.cpp (hunk 9 above)
Comment (ajreynol) [reply src/smt/smt_engine.cpp:84]: No, ill remove.  (There are many in this list we don't need actually) :)

File: src/smt/smt_engine.h (hunk 10 above)
Comment (ajreynol) [reply src/smt/smt_engine.h:513]: It is indeed a simple change, I'd prefer not writing any new Expr-level code, which is why I changed here.

File: src/smt/abduction_solver.h [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+   * If this method returns true, then abd is set to a formula C such that
+   * A ^ C is satisfiable, and A ^ ~B ^ C is unsatisfiable.
+   *
+   * The argument grammarType is a sygus datatype type that encodes the syntax
+   * restrictions on the shape of possible solutions.
+   *
+   * This method invokes a separate copy of the SMT engine for solving the
+   * corresponding sygus problem for generating such a solution.
+   */
+  bool getAbduct(const Node& conj, const TypeNode& grammarType, Node& abd);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/abduction_solver.h:53]: Done.

File: src/smt/abduction_solver.h [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+  /**
+   * This method asks this SMT engine to find an abduct with respect to the
+   * current assertion stack (call it A) and the goal (call it B).
+   * If this method returns true, then abd is set to a formula C such that
+   * A ^ C is satisfiable, and A ^ ~B ^ C is unsatisfiable.
+   *
+   * @param goal The goal of the abduction problem.
+   * @param grammarType A sygus datatype type that encodes the syntax
+   * restrictions on the shape of possible solutions.
+   * @param abd This argument is update to contain the solution to the
```
Comment (yoni206) [inline src/smt/abduction_solver.h:50]:
```suggestion
   * @param abd This argument is updated to contain the solution to the
```
